### PR TITLE
go/runtime/host/sgx: Add metrics for attestations

### PR DESCRIPTION
**Pull request by abukosek**
This PR adds three new metrics relating to SGX TEE attestations:
- `oasis_tee_attestations_performed` - Number of TEE attestations performed.
- `oasis_tee_attestations_successful` - Number of successful TEE attestations.
- `oasis_tee_attestations_failed` - Number of failed TEE attestations.

### CODE REVIEW

File: go/runtime/host/sgx/sgx.go [hunk 1]
```diff
@@ -44,6 +47,43 @@ const (
 	defaultRuntimeAttestInterval = 1 * time.Hour
 )
 
+var (
```
Comment (kostko) [inline go/runtime/host/sgx/sgx.go:50]: Could we move these to a separate `metrics.go`?

File: go/runtime/host/sgx/sgx.go [hunk 2]
```diff
@@ -136,7 +176,19 @@ func (ts *teeState) update(ctx context.Context, sp *sgxProvisioner, conn protoco
 	if ts.impl == nil {
 		return nil, fmt.Errorf("not initialized")
 	}
-	return ts.impl.Update(ctx, sp, conn, report, nonce)
+
+	retB, retE := ts.impl.Update(ctx, sp, conn, report, nonce)
+
+	if metrics.Enabled() {
```
Comment (kostko) [inline go/runtime/host/sgx/sgx.go:182]: No need to gate these, just always update the metrics.

File: go/runtime/host/sgx/sgx.go [hunk 3]
```diff
@@ -136,7 +141,19 @@ func (ts *teeState) update(ctx context.Context, sp *sgxProvisioner, conn protoco
...
+		teeAttestationsPerformed.With(prometheus.Labels{"runtime": ts.runtimeID.String()}).Inc()
+		if retE != nil {
+			teeAttestationsFailed.With(prometheus.Labels{"runtime": ts.runtimeID.String(), "error": retE.Error()}).Inc()
```
Comment (peternose) [inline go/runtime/host/sgx/sgx.go:150]: Do we need the error message or is `status=succeeded|failed` enough? If the error is unique on every failure (e.g. it contains a timestamp), we will create a lot of times series here.

File: go/runtime/host/sgx/sgx.go [hunk 4]
```diff
@@ -421,6 +438,12 @@ func New(cfg Config) (host.Provisioner, error) {
 		cfg.RuntimeAttestInterval = defaultRuntimeAttestInterval
 	}
 
+	if metrics.Enabled() {
```
Comment (peternose) [inline go/runtime/host/sgx/sgx.go:441]: We could also move this to `metrics.go` (e.g. create a new function `initMetrics`).

File: go/runtime/host/sgx/sgx.go [hunk 5]
```diff
@@ -136,7 +141,19 @@ func (ts *teeState) update(ctx context.Context, sp *sgxProvisioner, conn protoco
 	if ts.impl == nil {
 		return nil, fmt.Errorf("not initialized")
 	}
-	return ts.impl.Update(ctx, sp, conn, report, nonce)
+
+	retB, retE := ts.impl.Update(ctx, sp, conn, report, nonce)
+
+	if metrics.Enabled() {
```
[the remark this answers is not in the document]
Comment (abukosek) [reply go/runtime/host/sgx/sgx.go:147]: Hmm, why?  Elsewhere, we only update metrics if metrics are actually enabled.

File: go/runtime/host/sgx/metrics.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package sgx
+
+import "github.com/prometheus/client_golang/prometheus"
+
+var (
+	// Number of TEE attestations performed.
+	teeAttestationsPerformed = prometheus.NewCounterVec(
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/runtime/host/sgx/metrics.go:7]: Yeah let's leave it as 3 separate metrics.

File: go/runtime/host/sgx/sgx.go (hunk 3 above)
Comment (kostko) [reply go/runtime/host/sgx/sgx.go:150]: I agree that the error message is too much. If one needs the actual error message, it should be looked up in the node's logs.

Comment (abukosek) [reply go/runtime/host/sgx/sgx.go]: OK, I'll remove it :)

**Review by peternose — APPROVED** [review]
Tested and works :+1: 
```
# HELP oasis_tee_attestations_performed Number of TEE attestations performed.
# TYPE oasis_tee_attestations_performed counter
oasis_tee_attestations_performed{runtime="c000000000000000ffffffffffffffffffffffffffffffffffffffffffffffff"} 3
# HELP oasis_tee_attestations_successful Number of successful TEE attestations.
# TYPE oasis_tee_attestations_successful counter
oasis_tee_attestations_successful{runtime="c000000000000000ffffffffffffffffffffffffffffffffffffffffffffffff"} 3
```

**Comment by abukosek** [comment]
Thanks! :D

**Pull request merged**